--- a/chrome/browser/ui/browser_ui_prefs.cc	2018-06-21 21:25:48.222493615 -0400
+++ b/chrome/browser/ui/browser_ui_prefs.cc.new	2018-06-21 21:26:37.892817082 -0400
@@ -74,7 +74,7 @@
       user_prefs::PrefRegistrySyncable::SYNCABLE_PREF);
   registry->RegisterStringPref(prefs::kCloudPrintEmail, std::string());
   registry->RegisterBooleanPref(prefs::kCloudPrintProxyEnabled, true);
-  registry->RegisterBooleanPref(prefs::kCloudPrintSubmitEnabled, true);
+  registry->RegisterBooleanPref(prefs::kCloudPrintSubmitEnabled, false);
   registry->RegisterBooleanPref(prefs::kDevToolsDisabled, false);
   registry->RegisterDictionaryPref(prefs::kBrowserWindowPlacement);
   registry->RegisterDictionaryPref(prefs::kBrowserWindowPlacementPopup);
@@ -89,10 +89,10 @@
 #endif
 #if BUILDFLAG(ENABLE_WEBRTC)
   // TODO(guoweis): Remove next 2 options at M50.
-  registry->RegisterBooleanPref(prefs::kWebRTCMultipleRoutesEnabled, true);
-  registry->RegisterBooleanPref(prefs::kWebRTCNonProxiedUdpEnabled, true);
+  registry->RegisterBooleanPref(prefs::kWebRTCMultipleRoutesEnabled, false);
+  registry->RegisterBooleanPref(prefs::kWebRTCNonProxiedUdpEnabled, false);
   registry->RegisterStringPref(prefs::kWebRTCIPHandlingPolicy,
-                               content::kWebRTCIPHandlingDefault);
+                               content::kWebRTCIPHandlingDisableNonProxiedUdp);
   registry->RegisterStringPref(prefs::kWebRTCUDPPortRange, std::string());
 #endif
 
