--- a/third_party/blink/renderer/platform/scheduler/utility/webthread_impl_for_utility_thread.cc	
2018-08-16 03:09:14.000000000 -0400
+++ b/third_party/blink/renderer/platform/scheduler/utility/webthread_impl_for_utility_thread.cc.new	
2018-08-16 23:27:38.273887536 -0400
@@ -16,7 +16,6 @@
 WebThreadImplForUtilityThread::~WebThreadImplForUtilityThread() = default;
 
 blink::ThreadScheduler* WebThreadImplForUtilityThread::Scheduler() const {
-  NOTIMPLEMENTED();
   return nullptr;
 }
 
